moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp

```diff
@@ -874,83 +880,342 @@ bool TimeOptimalTrajectoryGeneration::computeTimeStamps(robot_trajectory::RobotT
           std::min(std::fabs(bounds.max_velocity_), std::fabs(bounds.min_velocity_)) * velocity_scaling_factor;
+      set_velocity_limit = true;
     }
 
-    max_acceleration[j] = 1.0;
-    if (bounds.acceleration_bounded_)
+    if (!set_velocity_limit)
+    {
+      ROS_ERROR_STREAM_NAMED(LOGNAME, "No velocity limit was defined for joint "
+                                          << vars[idx].c_str()
+                                          << "! You have to define velocity limits in the URDF or "
+                                             "joint_limits.yaml");
+      return false;
+    }
+
+    // ACCELERATION LIMIT
+    // Check if a custom limit was specified as an argument
+    bool set_acceleration_limit = false;
+    it = acceleration_limits.find(vars[idx]);
+    if (it != acceleration_limits.end())
     {
-      if (bounds.max_acceleration_ < std::numeric_limits<double>::epsilon())
+      max_acceleration[idx] = it->second * acceleration_scaling_factor;
+      set_acceleration_limit = true;
+    }
+
+    if (bounds.acceleration_bounded_ && !set_acceleration_limit)
+    {
+      // Set the default acceleration limit, from robot model
+      if (bounds.max_acceleration_ < 0.0)
       {
         ROS_ERROR_NAMED(LOGNAME, "Invalid max_acceleration %f specified for '%s', must be greater than 0.0",
-                        bounds.max_acceleration_, vars[j].c_str());
+                        bounds.max_acceleration_, vars[idx].c_str());
         return false;
       }
-      max_acceleration[j] = std::min(std::fabs(bounds.max_acceleration_), std::fabs(bounds.min_acceleration_)) *
-                            acceleration_scaling_factor;
+      max_acceleration[idx] = std::min(std::fabs(bounds.max_acceleration_), std::fabs(bounds.min_acceleration_)) *
+                              acceleration_scaling_factor;
+      set_acceleration_limit = true;
+    }
+    if (!set_acceleration_limit)
+    {
+      ROS_ERROR_STREAM_NAMED(LOGNAME, "No acceleration limit was defined for joint "
+                                          << vars[idx].c_str()
+                                          << "! You have to define acceleration limits in the URDF or "
+                                             "joint_limits.yaml");
+      return false;
     }
   }
 
+  return doTimeParameterizationCalculations(trajectory, max_velocity, max_acceleration);
+}
+
+bool TimeOptimalTrajectoryGeneration::computeTimeStampsWithTorqueLimits(
+    robot_trajectory::RobotTrajectory& trajectory, const geometry_msgs::Vector3& gravity_vector,
+    const std::vector<double>& joint_torque_limits, double accel_limit_decrement_factor,
+    const std::unordered_map<std::string, double>& velocity_limits,
+    const std::unordered_map<std::string, double>& acceleration_limits, const double max_velocity_scaling_factor,
+    const double max_acceleration_scaling_factor) const
+{
+  // 1. Call computeTimeStamps() to time-parameterize the trajectory with given vel/accel limits.
```

Comment:
Is there any reason why this should be associated with TOTG? To me this looks like it could be a stand-alone trajectory processing feature.

Reply:
Well, it is based on TOTG. See L1116.

But making it a standalone feature seems OK too. Possibly keep it in the `trajectory_processing` folder, call it `time_optimal_with_torque_limits`? Or `iterative_torque_limit_parameterization`?

Reply:
done